### ISSUE TITLE
Support UFO3

### ISSUE THREAD

**Issue by miguelsousa** [adobe-type-tools/psautohint#23]

**Comment by khaledhosny** [comment]
That was part of the motivation of #7 and #8, may be we should revisit them again.

**Comment by miguelsousa** [comment]
Yeah I'd like to dig into it. Support to UFO3 is something we need definitely. Not sure if the best long-term approach is to build on top of the existing modules or to try to leverage external pieces.

For starters, can you state what are the requirements of psautohint? For example, can it only consume data in bez format?

**Comment by khaledhosny** [comment]
The C library only reads and writes bez format, everything else is handled by the Python code.

Reading UFO files is done by `ufoFont.py` which parses the XML files directly. It might not be hard to extend it (I didn’t check closely), but I had my eye on future UFO updates like https://github.com/unified-font-object/ufoLib/pull/26.

**Comment by madig** [comment]
I was wondering why you might want to hint a UFO as opposed to just provide blue values and then run psautohint on the OTFs?

**Comment by miguelsousa** [comment]
> I was wondering why you might want to hint a UFO as opposed to just provide blue values and then run psautohint on the OTFs?

Good question. I think the reason is mostly due to carrying over our traditional workflow that was based on using PFA sources. However, it's also due to a limitation in the toolchain; some years ago I had the same though as you and was made aware that although the OTF that autohint/psautohint returns is not optimized (i.e. the charstrings are not specialized nor subroutinized).

Also, there are situations where one may want to keep the hints under version control.

**Comment by madig** [comment]
Huh. So makeotf optimizes the OTF's hinting?

**Comment by miguelsousa** [comment]
makeotfexe does the optimization of the CFF table.

**Comment by madig** [comment]
But what does that have to do with psautohint and optimization?

**Comment by miguelsousa** [comment]
In order to add the hinting operators to the CFF's charstrings, psautohint de-optimizes the table first (and does not re-optimize it at the end).

**Comment by anthrotype** [comment]
So if we want to add it to ufo2ft we would have to first write a non-optimised CFF then run psautohint on it, then optimise it with fonttools and then compreff it

**Comment by anthrotype** [comment]
Or psautohint could do the CFF charstrings specialisation using fonttools (since it’s already a requirement)

**Comment by miguelsousa** [comment]
> Or psautohint could do the CFF charstrings specialisation using fonttools

yes, that should be the route, I think. Remind me, can fonttools do subroutinization as well?

**Comment by khaledhosny** [comment]
I started porting `ufoFont.py` to ufoLib in [wip/ufolib](https://github.com/adobe-type-tools/psautohint/tree/wip/ufolib) branch, reading should be completely ported to ufoLib and a bit of writing. I’m now feeling this is a bit too low level, and all that two layer setup when processing UFO fonts is making things needlessly complex. I need to understand the UFO workflow better and see if things can be simplified, so I’m putting this on hold for now. Just in case anyone wants to have a look.

**Comment by miguelsousa** [comment]
Sounds good, thanks
